pkg/networking/cilium/templater.go

```diff
@@ -43,57 +44,81 @@ func (c *Templater) GenerateUpgradePreflightManifest(ctx context.Context, spec *
-	currentVersion, err := semver.New(currentSpec.VersionsBundle.Cilium.Version)
-	if err != nil {
-		return nil, fmt.Errorf("invalid version for Cilium in current spec: %v", err)
-	}
-
-	v := templateValues(newSpec)
-	v.set(fmt.Sprintf("%d.%d", currentVersion.Major, currentVersion.Minor), "upgradeCompatibility")
+// ManifestOpt allows to modify options for a cilium manifest
+type ManifestOpt func(*ManifestConfig)
 
-	uri, version := getChartUriAndVersion(newSpec)
+type ManifestConfig struct {
+	values      values
+	kubeVersion string
+	namespaces  []string
+}
 
-	kubeVersion, err := getKubeVersionString(currentSpec)
-	if err != nil {
-		return nil, err
+// WithKubeVersion allows to generate the Cilium manifest for a different kubernetes version
+// than the one specified in the cluster spec. Useful for upgrades scenarios where Cilium is upgraded before
+// the kubernetes components
+func WithKubeVersion(kubeVersion string) ManifestOpt {
+	return func(c *ManifestConfig) {
+		c.kubeVersion = kubeVersion
 	}
+}
 
-	manifest, err := c.helm.Template(ctx, uri, version, namespace, v, kubeVersion)
-	if err != nil {
-		return nil, fmt.Errorf("failed generating cilium upgrade manifest: %v", err)
+// WithUpgradeFromVersion allows to specify the compatibility Cilium version to use in the manifest.
+// This is necessary for Cilium upgrades
+func WithUpgradeFromVersion(version semver.Version) ManifestOpt {
+	return func(c *ManifestConfig) {
+		c.values.set(fmt.Sprintf("%d.%d", version.Major, version.Minor), "upgradeCompatibility")
 	}
-
-	return manifest, nil
 }
 
-func (c *Templater) GenerateManifest(ctx context.Context, spec *cluster.Spec) ([]byte, error) {
-	v := templateValues(spec)
-
-	uri, version := getChartUriAndVersion(spec)
+// WithPolicyAllowedNamespaces allows to specify which namespaces traffic should be allowed when using
+// and "Always" policy enforcement mode
+func WithPolicyAllowedNamespaces(namespaces []string) ManifestOpt {
+	return func(c *ManifestConfig) {
+		c.namespaces = namespaces
+	}
+}
 
+func (t *Templater) GenerateManifest(ctx context.Context, spec *cluster.Spec, opts ...ManifestOpt) ([]byte, error) {
 	kubeVersion, err := getKubeVersionString(spec)
 	if err != nil {
 		return nil, err
 	}
 
-	manifest, err := c.helm.Template(ctx, uri, version, namespace, v, kubeVersion)
+	c := &ManifestConfig{
+		values:      templateValues(spec),
+		kubeVersion: kubeVersion,
+	}
+	for _, o := range opts {
+		o(c)
+	}
+
+	uri, version := getChartUriAndVersion(spec)
+
+	manifest, err := t.helm.Template(ctx, uri, version, namespace, c.values, c.kubeVersion)
 	if err != nil {
 		return nil, fmt.Errorf("failed generating cilium manifest: %v", err)
 	}
 
+	if spec.Cluster.Spec.ClusterNetwork.CNIConfig.Cilium.PolicyEnforcementMode == anywherev1.CiliumPolicyModeAlways {
+		networkPolicyManifest, err := t.GenerateNetworkPolicyManifest(spec, c.namespaces)
```

Comment:
Since we only set `c.namespaces` in `WithPolicyAllowedNamespaces` now, what would be the behavior if there are no namespaces? 
Alternatively, if there will always be namespaces here, what's the use case for making this a ManifestOpt?

Reply:
The behavior is that no traffic in extra namespaces will be allowed with network policies.

The controller will use this without the namespaces.